### ISSUE TITLE
Add Debug draw helpers for ShapeCast & CastRay

### ISSUE THREAD

**Issue by Joshua-Ashton** [jrouwe/JoltPhysics#96]
It would be really helpful for debugging to have the ability to ->Draw a `ShapeCast` or `RayCast` wiuth the same params as a CastShapeVsShape or CastRay.

Obviously one can do their own with pShape->Draw, but for debugging sanity it would be really helpful to have a simple way to visualize individual traces and rays going on without any potential mess ups in the app's code getting the CentreOfMass offsets wrong, rotations/inverse rotations etc.

Right now in our code, that is what we are doing, with a little picture in picture window showing all the traces, but obviously it would be super nice to validate against exactly what Jolt is seeing.

**Comment by jrouwe** [comment]
I'm not sure if I understand correctly. Drawing a ShapeCast or RayCast would be easy (it's only a couple of lines of code, see e.g. SamplesApp::CastProbe), but since there are various ways you can use that Ray/ShapeCast that work in different spaces it may be just as confusing. E.g. Shape::RayCast works against the shape centered around its COM,  TransformedShape::CastRay / NarrowPhaseQuery::CastRay takes a world space RayCast.

**Comment by Joshua-Ashton** [comment]
I guess this stems from what we are trying to debug with our current integration.

Right now, when we take our ShapeCast and set the `inCenterOfMassStart` to that of the worldspace box center and `inDirection` to it's direction, and then set `inCenterOfMassTransform2` to the worldspace transform of the object, we get no hits except at the origin of the map. Whereas if we instead take the ShapeCast and PostTransform it with the inverse of the transform of the object and set the `inCenterOfMassTransform2` to Identity, things hit fine, and we were wanting to visualize why this is happening cause it doesn't seem to make much sense to me at least.

A secondary thing we are trying to debug is whether the output normal after rotating it is correct, and also the two objects.
Right now there's an issue where we are getting hits that we are reporting as starting inside, whereas on the game side they should not be a hit, it should just be 'touching' ie. frac = 0 but not a 'hit'. And we were wanting to see visually what was happening here to try and resolve it (eg. take the dot product of the normal post-rotation and the ray dir to modify the real hit length and stuff to see if we can make this behave like we had in the old physics system or whatever.)

Sorry for the kind of vague wall of text here.

We have some stuff to visualize right now, but we were wanting to make sure that we weren't completely off-base looking at the wrong thing so I was thinking maybe it would be better if Jolt had some helpers to debug here so we can see *exactly* what Jolt is seeing wrt the two objects on its end.

**Comment by jrouwe** [comment]
My guess is that the object that you're colliding against has a Shape::GetCenterOfMass() of non-zero. If you pass in the world transform of that object, you're not passing in the center of mass transform (which is what inCenterOfMassTransform2 expects). You should be passing inCenterOfMassTransform2 = WorldTransform * Mat44::sTranslation(target_shape->GetCenterOfMass()).

**Comment by jrouwe** [comment]
Hmmm, it looks like your code is correct (or at least the commented and uncommented parts are equivalent as far as I can see). I would need to debug this. Can you provide some code to initialize origin, direction, position, rotation, ray.m_Extents and pShape for a case where the commented code works and the uncommented code doesn't? (I can insert my own filter and collector)

**Comment by Joshua-Ashton** [comment]
Sure, hopefully I can get away with serializing my shapes and unserializing for the test

**Comment by Joshua-Ashton** [comment]
Hi, I attempted to make that test today, but unfortunately serializing the shape doesn't work correctly, it is a `StaticCompoundShape` and all of the subshapes become NULL when restoring from the binary

**Comment by Joshua-Ashton** [comment]
Here is the broken shape:

[my_collide.bin.removethetxt.txt](https://github.com/jrouwe/JoltPhysics/files/8415001/my_collide.bin.removethetxt.txt)

**Comment by Joshua-Ashton** [comment]
Ah, I think I need to use SaveSubShapeState and such...

**Comment by Joshua-Ashton** [comment]
Welp, that doesn't work, just fails with "Failed to restore shape."

Here is the .bin for the shape to use with `sRestoreWithChildren` that fails and was dumped straight from the engine with `SaveWithChildren` straight to a file.

[my_collide_new.bin.removethetxt.txt](https://github.com/jrouwe/JoltPhysics/files/8415045/my_collide_new.bin.removethetxt.txt)

**Comment by jrouwe** [comment]
Do you also have all other parameters (position etc.)? I.e can you write a code snippet that works with the commented out code but doesn't with the other code?

**Comment by Joshua-Ashton** [comment]
Sure:

```cpp
		std::ifstream myfile("C:\\Users\\Joshua\\my_collide.bin");
		JPH::StreamInWrapper wrapper(myfile);
		JPH::Shape::IDToShapeMap id_to_shape;
		JPH::Shape::IDToMaterialMap id_to_material;
		auto shapey = JPH::Shape::sRestoreWithChildren(wrapper, id_to_shape, id_to_material).Get();

		JPH::Shape* pShape = shapey.GetPtr();

		JPH::Vec3 collidePosition = Vec3(0x1.00d8440000000p+5, 0x1.1e1b080000000p+2, 0x0.0000000000000p+0);
		JPH::Quat collideRotation = Quat(0x0.0000000000000p+0, 0x0.0000000000000p+0, 0x0.0000000000000p+0, 0x1.fffffc0000000p-1);

		JPH::Mat44 queryTransform = JPH::Mat44::sRotationTranslation(collideRotation, collidePosition + collideRotation * pShape->GetCenterOfMass());

		JPH::Vec3 boxPosition = JPH::Vec3(0x1.f84b940000000p+4, 0x1.81e37e0000000p+1, 0x1.dab9d20000000p+2);
		JPH::Vec3 boxDirection = JPH::Vec3(0x0.0000000000000p+0, 0x0.0000000000000p+0, -0x1.0418940000000p-1);
		JPH::Vec3 boxHalfExtent = JPH::Vec3(0x1.a027520000000p-2, 0x1.a027520000000p-2, 0x1.d42c3c0000000p-1);

		JPH::BoxShape boxShape(boxHalfExtent);

		JPH::ShapeCast initialShapeCast(&boxShape, JPH::Vec3::sReplicate(1.0f), JPH::Mat44::sTranslation(boxPosition), boxDirection);
		//JPH::ShapeCast shapeCast = initialShapeCast.PostTransformed(queryTransform.InversedRotationTranslation());
		JPH::ShapeCast shapeCast = initialShapeCast;

		JPH::ShapeCastSettings settings;
		JPH::ShapeFilter filter;
		JPH::AnyHitCollisionCollector<JPH::CastShapeCollector> collector;
		//JPH::CollisionDispatch::sCastShapeVsShape(shapeCast, settings, pShape, JPH::Vec3::sReplicate(1.0f), filter, JPH::Mat44::sIdentity(), JPH::SubShapeIDCreator(), JPH::SubShapeIDCreator(), collector);
		JPH::CollisionDispatch::sCastShapeVsShape(shapeCast, settings, pShape, JPH::Vec3::sReplicate(1.0f), filter, queryTransform, JPH::SubShapeIDCreator(), JPH::SubShapeIDCreator(), collector);

		if (!collector.HadHit())
		{
			printf("Test failed, womp womp\n");
		}
		else
		{
			printf("Test passed, woo!\n");
		}
```

**Comment by jrouwe** [comment]
Ok, I can reproduce. The reason your shape wasn't deserializing is because you opened the file as a text file. Should have done:

`std::ifstream myfile("my_collide.bin", ifstream::in | ifstream::binary);`

Now onto finding out what goes wrong.

**Comment by jrouwe** [comment]
Ok, I should have seen this right away.

The issue is that sCastShapeVsShape expects inShapeCast to be relative to the center of mass of the shape that you cast against. So queryTransform is the center of mass transform of the shape you cast against, initialShapeCast contains the shape cast in world space (note that if your cast shape was not a box and had a center of mass offset then you would need to correct for it in mCenterOfMassStart), so the correct thing to do is indeed to transform initialShapeCast with the inverse of queryTransform to get it in the center of mass space:

`JPH::ShapeCast shapeCast = initialShapeCast.PostTransformed(queryTransform.InversedRotationTranslation());`

sCastShapeVsShape's inCenterOfMassTransform2 is only used to transform the hit results back into world space. So if you do:

`JPH::CollisionDispatch::sCastShapeVsShape(shapeCast, settings, pShape, JPH::Vec3::sReplicate(1.0f), filter, JPH::Mat44::sIdentity(), JPH::SubShapeIDCreator(), JPH::SubShapeIDCreator(), collector);`

then you're going to transform the hit point and normal by identity. So instead you should do:

`JPH::CollisionDispatch::sCastShapeVsShape(shapeCast, settings, pShape, JPH::Vec3::sReplicate(1.0f), filter, queryTransform, JPH::SubShapeIDCreator(), JPH::SubShapeIDCreator(), collector);`

so that the hit results passed to the collector will be in world space.

I'm going to see if I can make a helper function for this because the interface is a bit confusing.

**Comment by Joshua-Ashton** [comment]
I see, we were manually rotating/translating those after, seems weird for the func to work like this when you can seemingly do it yourself.

Also yeah, ifstream moment...

**Comment by jrouwe** [comment]
The sCastShapeVsShape is used recursively (so a compound shape calls it on its children). If we don't keep track of the current COM transform, you will get your hit results in the space of a random child and it will be difficult to get everything back in world space (it's not a matter of multiplying with queryTransform because then you miss the transform of the child). 

Also CastShapeSettings::mActiveEdgeMovementDirection wouldn't work without knowing what the current COM transform is.

I've created #116 to make shape casting a little bit easier.

**Comment by Joshua-Ashton** [comment]
Thanks a lot! Will try it out later today.

Really appreciate all you have done so far!

**Comment by jrouwe** [comment]
I've merged #116 into master. Closing the ticket (reopen if you need anything else).
